src/blokus.c

```diff
@@ -0,0 +1,515 @@
+            int y = pos.y + tile->blks[i].y, x = pos.x + tile->blks[i].x;
+
+            // Out of map
+            if (!within_map(y, x))
+                return -1;
+
+            // Check if covering starting point
+            if (y == STARTING_POINT[p].y && x == STARTING_POINT[p].x)
+                valid = 1;
+        }
+    } else {
+        for (int i = 0; i < TILE[tile->shape].blk_cnt; ++i) {
+            int y = pos.y + tile->blks[i].y, x = pos.x + tile->blks[i].x;
+
+            // Out of map or overlapping
+            if (!within_map(y, x) || gcb->map[y][x] != -1)
+                return -1;
+
+            for (int j = 0; j < 4; ++j) {
+                int ey = y + EDGE[j].y, ex = x + EDGE[j].x;
+
+                // Edge-to-edge contact is not allowed
+                if (within_map(ey, ex) && gcb->map[ey][ex] == p)
+                    return -1;
+            }
+
+            for (int j = 0; j < 4; ++j) {
+                int cy = y + CORNER[j].y, cx = x + CORNER[j].x;
+
+                // Must have at least one corner-to-corner contact
+                if (within_map(cy, cx) && gcb->map[cy][cx] == p)
+                    valid = 1;
+            }
+        }
+    }
+
+    return (valid)? 0 : -1;
+}
+
+gcb_t *init_gcb(int turn)
+{
+    gcb_t *gcb = malloc(sizeof(gcb_t));
+    gcb->turn = turn;
+    gcb->sel_shape = -1;
+    gcb->score[0] = 0;
+    gcb->score[1] = 0;
+    for (int s = 0; s <= SHAPE_Z; ++s) {
+        gcb->hand[0][s] = make_tile(s);
+        gcb->hand[1][s] = make_tile(s);
+    }
+    for (int y = 0; y < N_ROW; ++y) {
+        for (int x = 0; x < N_COL; ++x) {
+            gcb->map[y][x] = -1;
+        }
+    }
+    for (int i = 0; i < 195; ++i)
+        gcb->next_empty[i] = i + 1;
+    gcb->next_empty[195] = -1;
+    for (int i = 1; i < 196; ++i)
+        gcb->prev_empty[i] = i - 1;
+    gcb->prev_empty[0] = -1;
+    return gcb;
+}
+
+tile_t *sel_tile(gcb_t *gcb, int shape)
+{
+    gcb->sel_shape = shape;
+    return gcb->hand[gcb->turn][gcb->sel_shape];
+}
+
+int can_place(gcb_t *gcb)
+{
+    int p = gcb->turn;
+    tile_t *tile;
+    if (gcb->sel_shape == -1)
+        return -1;
+
+    tile = gcb->hand[p][gcb->sel_shape];
+    return test_place(gcb, tile) == 0;
+}
+
+int update(gcb_t *gcb, char *code)
+{
+    int p = gcb->turn;
+    tile_t *tile = code ? decode_tile(code) : gcb->hand[p][gcb->sel_shape];
+
+    if (test_place(gcb, tile) < 0)
+        return -1;  // reject invalid update
+
+    coord_t pos = tile->pos;
+    for (int i = 0; i < TILE[tile->shape].blk_cnt; ++i) {
+        int y = pos.y + tile->blks[i].y, x = pos.x + tile->blks[i].x;
+        int k = N_ROW * y + x;
+        gcb->map[y][x] = p;
+        gcb->next_empty[gcb->prev_empty[k]] = gcb->next_empty[k];
+        gcb->prev_empty[gcb->next_empty[k]] = gcb->prev_empty[k];
+    }
+
+    encode_tile(gcb->code, tile);  // record the last update with gcb->code
+    gcb->score[p] += TILE[tile->shape].blk_cnt;
```

Comment:
We should set the tile on user hand to `NULL` if they place the tile correctly

Reply:
OK! Please wait for my update.